### ISSUE TITLE
Issue when running a module inside pex file

### ISSUE THREAD

**Issue by silverguo** [pantsbuild/pex#1018]
I met some strange behaviors when running a module inside pex file. I found a simple reproducible example as below.

Python: 3.8.3
Pex: 2.1.15

I have a `test.py` under a folder `src`,
```
[cwd]
└── src
    └── test.py
```

`test.py`:
```python
import pickle

def add(a: int, b: int) -> int:
    return a+b

def main() -> None:

    pickle_add = pickle.dumps(add)
    add_clone = pickle.loads(pickle_add)
    print(add_clone(1, 2))

if __name__ == "__main__":
    main()
```

To build a pex file with entry point:
    
    pex -D ./src -e test:main -o test_with_ep.pex

To build a pex file without entry point:

    pex -D ./src -o test_no_ep.pex

No issue when running:

    python test_with_ep.pex      
    3

But when running `test_no_ep.pex`:

    python test_no_ep.pex -m test
    Traceback (most recent call last):
      File "/private/tmp/test_no_ep.pex/.bootstrap/pex/pex.py", line 423, in execute
      File "/private/tmp/test_no_ep.pex/.bootstrap/pex/pex.py", line 356, in _wrap_coverage
      File "/private/tmp/test_no_ep.pex/.bootstrap/pex/pex.py", line 387, in _wrap_profiling
      File "/private/tmp/test_no_ep.pex/.bootstrap/pex/pex.py", line 477, in _execute
      File "/private/tmp/test_no_ep.pex/.bootstrap/pex/pex.py", line 520, in execute_interpreter
      File "/private/tmp/test_no_ep.pex/.bootstrap/pex/pex.py", line 584, in execute_module
      File "/Users/yuhan/.pyenv/versions/3.8.3/lib/python3.8/runpy.py", line 210, in run_module
        return _run_code(code, {}, init_globals, run_name, mod_spec)
      File "/Users/yuhan/.pyenv/versions/3.8.3/lib/python3.8/runpy.py", line 87, in _run_code
        exec(code, run_globals)
      File "/private/tmp/test_no_ep.pex/test.py", line 13, in <module>
      File "/private/tmp/test_no_ep.pex/test.py", line 8, in main
    _pickle.PicklingError: Can't pickle <function add at 0x1042aef70>: attribute lookup add on __main__ failed

Problem could be solved if I move that function `main()` into another file and import it in `test.py`. Then I could run that module in pex file.

Do I miss something here ?

**Comment by benjyw** [comment]
Hmm, well, I can reproduce this at least, so it's consistent. Will take a look.

**Comment by benjyw** [comment]
This looks like a quirk of how pickle works with module-level objects, possibly related to https://bugs.python.org/issue37852.

**Comment by jsirois** [comment]
It looks like the issue here is Pex use of `runpy.run_module`. Passing `alter_sys=True`, which is what `python -m` does, solves this.

See the CPython main -m handling here which uses `runpy.run_module`:
https://github.com/python/cpython/blob/ba251c2ae6654bfc8abd9d886b219698ad34ac3c/Modules/main.c#L580-L581
https://github.com/python/cpython/blob/ba251c2ae6654bfc8abd9d886b219698ad34ac3c/Modules/main.c#L258-L283

**Comment by benjyw** [comment]
Presumably this would not affect pexes built with --venv?

**Comment by jsirois** [comment]
> Presumably this would not affect pexes built with --venv?

Nope - it also affects --venv mode. Both normal and --venv still insert a `__main__` that delegates to the configured entry point and both delegations suffer the same issue.

**Comment by benjyw** [comment]
Aha, that is kind of good news as it might explain some weirdness I'm seeing when running mypy on its own code from a --venv pex 

It's a bit of a stretch though.

**Comment by jsirois** [comment]
Thanks for your report and patience @silverguo. The fix for this will go out in 2.1.34 today. I followed your example exactly in an integration test - EXCEPT - I had to rename the module from test.py to issues_1018.py. Who knew! Python 2, PyPy and PyPy3 - but not Python 3.5+ have a builtin test module and `python -m test` just starts running all the core python tests!
